src/simdvec.jl

```diff
@@ -0,0 +1,458 @@
+        Vec($(llvmop)(x.data, y.data))
+    end
+end
+
+# overflow
+
+const OVERFLOW_INTRINSICS = [
+    (:(Base.Checked.add_with_overflow) , IntTypes  , Intrinsics.sadd_with_overflow)
+    (:(Base.Checked.add_with_overflow) , UIntTypes , Intrinsics.uadd_with_overflow)
+    (:(Base.Checked.sub_with_overflow) , IntTypes  , Intrinsics.ssub_with_overflow)
+    (:(Base.Checked.sub_with_overflow) , UIntTypes , Intrinsics.usub_with_overflow)
+    (:(Base.Checked.mul_with_overflow) , IntTypes  , Intrinsics.smul_with_overflow)
+    (:(Base.Checked.mul_with_overflow) , UIntTypes , Intrinsics.umul_with_overflow)
+]
+for (op, constraint, llvmop) in OVERFLOW_INTRINSICS
+    @eval @inline function $op(x::Vec{N, T}, y::Vec{N, T}) where {N, T <: $constraint}
+        val, overflows = $(llvmop)(x.data, y.data)
+        return Vec(val), Vec(overflows)
+    end
+end
+
+
+# max min
+@inline Base.max(v1::Vec{N,T}, v2::Vec{N,T}) where {N,T<:IntegerTypes} =
+    Vec(vifelse(v1 >= v2, v1, v2))
+@inline Base.min(v1::Vec{N,T}, v2::Vec{N,T}) where {N,T<:IntegerTypes} =
+    Vec(vifelse(v1 >= v2, v2, v1))
+
+# Pow
+@inline Base.:^(x::Vec{N,T}, y::IntegerTypes) where {N,T<:FloatingTypes} =
+    Vec(Intrinsics.powi(x.data, y))
+# Do what Base does for HWNumber:
+@inline Base.literal_pow(::typeof(^), x::Vec, ::Val{0}) = one(typeof(x))
+@inline Base.literal_pow(::typeof(^), x::Vec, ::Val{1}) = x
+@inline Base.literal_pow(::typeof(^), x::Vec, ::Val{2}) = x*x
+@inline Base.literal_pow(::typeof(^), x::Vec, ::Val{3}) = x*x*x
+
+# Sign
+@inline Base.flipsign(v1::Vec{N,T}, v2::Vec{N,T}) where {N,T} =
+    vifelse(signbit(v2), -v1, v1)
+@inline Base.copysign(v1::Vec{N,T}, v2::Vec{N,T}) where {N,T<:IntTypes} =
+    vifelse(signbit(v2), -abs(v1), abs(v1))
+_signed(::Type{Float32}) = Int32
+_signed(::Type{Float64}) = Int64
+@inline Base.signbit(x::Vec{N, T}) where {N, T <:FloatingTypes} =
+    signbit(reinterpret(Vec{N, _signed(T)}, x))
+
+# Pointer arithmetic
+for op in (:+, :-)
+    @eval begin
+        # Cast pointer to Int and back
+        @inline Base.$op(x::Vec{N,Ptr{T}}, y::Vec{N,Ptr{T}}) where {N,T} =
+            convert(Vec{N, Ptr{T}}, ($(op)(convert(Vec{N, Int}, x), convert(Vec{N, Int}, y))))
+        @inline Base.$op(x::Vec{N,Ptr{T}}, y::Union{IntegerTypes}) where {N,T} = $(op)(x, Vec{N,Ptr{T}}(y))
+        @inline Base.$op(x::IntegerTypes, y::Union{Vec{N,Ptr{T}}}) where {N,T} = $(op)(y, x)
+
+        @inline Base.$op(x::Vec{N,<:IntegerTypes}, y::Ptr{T}) where {N,T} = $(op)(Vec{N,Ptr{T}}(x), Vec{N,Ptr{T}}(y))
+        @inline Base.$op(x::Ptr{T}, y::Vec{N,<:IntegerTypes}) where {N,T} = $(op)(y, x)
+    end
+end
+
+# Bitshifts
+# See https://github.com/JuliaLang/julia/blob/7426625b5c07b0d93110293246089a259a0a677d/src/intrinsics.cpp#L1179-L1196
+# Shifting with a value larger than the number of bits in the type is undefined behavior
+# so set to zero in those cases.
+@inline function shl_int(x::Vec{N, T1}, y::Vec{N, T2}) where {N, T1<:IntegerTypes, T2<:IntegerTypes}
```

Comment:
I think it would be much better to use llvm-semantics and document that they differ from julia semantics. People who use `SIMD.jl` care about speed, and can deal with the fact that shifts by more that width bits behave weirdly.

Reply:
Not "weirdly", it gives you a poison value which gives you undefined behavior in many cases. I don't think we should expose undefined behavior so easily from the `Vec` type. You can always just call `Intrinsics.shl`?

Reply:
We could also do the bitshifts modulo the size of the integer (which would match with e.g. rust). That would be cheaper since you can just `and` it.

Reply:
Yes, this is a good idea. OpenCL does the same.

Reply:
What do we do about negative bitshifts?

Reply:
Julia distinguishes between signed and unsigned bitshifts. For signed ones, it handles negative values as positive shifts in the opposite direction.

If we use the same function/operator name, we should use the same semantics. For full performance, people have to specify unsigned shift counts. If we think this is confusing, then we can reject signed shift counts.

Reply:
> If we use the same function/operator name, we should use the same semantics.

But in https://github.com/eschnett/SIMD.jl/pull/63#discussion_r385139094 we basically said that we would not keep the same semantics? In Julia, a shift larger than the bitsize does not wrap around but sets the value to zero which to me is a very different semantics.

Reply:
Right. I thought Julia was undefined here, and we'd only be tightening things. But that's LLVM's semantics, not Julia's.

Should we keep Julia's semantics, and add new functions (`shl` / `shr`) that re-interpret as unsigned and wrap around?

In many cases, the shift count will be known at compile time, and in this case, things will be efficient anyway.

Reply:
So, I did some more reading on https://software.intel.com/sites/landingpage/IntrinsicsGuide/#techs=AVX,AVX2&cats=Shift

For hare-brained reason _mm256_slli_epi64 et al have julia semantics, while non-vectorized shifts ignore the upper bits. LLVM often fails to properly optimize code that enforces julia/AVX semantics; but if we emit x86_64 semantics, we have a mismatch with the AVX instructions we actually want. ARM does something else, because why not.

I see your point that the poison-value variant is slightly too poisonous for the default. So I guess we should document that poisonous `shl` is preferable to `<<` if known safe, and consider it our offering to ill-thought out llvm semantics.

Reply:
I hereby volunteer you to write the doc string for these functions.